### PR TITLE
Remove `Trainer._device_type`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of https://github.com/PyTorchLightning/pytorch-lightning/issues/11449

### Does your PR introduce any breaking changes? If yes, please list them.

Removes the protected `_strategy_type`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -305,7 +306,7 @@ def gpus_metrics(self) -> Dict[str, float]:
         .. deprecated:: v1.5
             Will be removed in v1.7.
         """
-        if self.trainer._device_type == _AcceleratorType.GPU and self.log_gpu_memory:
+        if isinstance(self.trainer.accelerator, GPUAccelerator) and self.log_gpu_memory:
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:309]: Alternatively we can use the strategy's root device type

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1716,33 +1715,31 @@ def __setup_profiler(self) -> None:
         self.profiler.setup(stage=self.state.fn._setup_fn, local_rank=local_rank, log_dir=self.log_dir)
 
     def _log_device_info(self) -> None:
-        rank_zero_info(f"GPU available: {torch.cuda.is_available()}, used: {self._device_type == _AcceleratorType.GPU}")
+        rank_zero_info(
+            f"GPU available: {torch.cuda.is_available()}, used: {isinstance(self.accelerator, GPUAccelerator)}"
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:1719]: Same here

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/loops/optimization/optimizer_loop.py [hunk 3]
```diff
@@ -369,7 +369,7 @@ def _optimizer_step(
             optimizer,
             opt_idx,
             train_step_and_backward_closure,
-            on_tpu=(self.trainer._device_type == _AcceleratorType.TPU and _TPU_AVAILABLE),
+            on_tpu=isinstance(self.trainer.accelerator, TPUAccelerator),
```
Comment (four4fish) [inline pytorch_lightning/loops/optimization/optimizer_loop.py:372]: We still need the _TPU_AVAILABLE check here. We haven't enable accelerator.is_available check yet. It's one of the follow up items. A lot of tests fails when we enable it, so didn't include in the big PR

**Review by four4fish — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:309]: I would prefer to stick with the Accelerators, as this one really is what defines the accelerator being used. For some accelerators it might make sense to have the `root_device` as cpu and then distribute it later within the strategy or something similar.

**Review by justusschock — APPROVED** [review]

**Comment by kaushikb11** [comment]
We have a logging issue right now on master.

```python
GPU available: False, used: False
TPU available: True, using: 8 TPU cores
IPU available: False, using: 0 IPUs
/usr/local/lib/python3.8/dist-packages/pytorch_lightning-1.6.0.dev0-py3.8.egg/pytorch_lightning/trainer/trainer.py:1737: UserWarning: TPU available but not used. Set the `tpu_cores` flag in your trainer `Trainer(tpu_cores=8)` or script `--tpu_cores=8`.
```
This PR should resolve it.

File: pytorch_lightning/loops/optimization/optimizer_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:372]: When would `TPUAccelerator` be selected without TPU available?

Comment (awaelchli) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:372]: yeah, self.trainer.accelerator = TPUAccelerator implies that _TPU_AVAILABLE, otherwise the acc connector wouldn't select it in the first place.

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -1173,29 +1174,31 @@ def val_dataloader(self):
...
+        ({"accelerator": "dp", "gpus": 2}, DataParallelStrategy, "dp", GPUAccelerator, 2),
+        ({"accelerator": "ddp", "gpus": 2}, DDPStrategy, "ddp", GPUAccelerator, 2),
+        ({"accelerator": "ddp2", "gpus": 2}, DDP2Strategy, "ddp2", GPUAccelerator, 2),
+        ({"accelerator": "ddp2", "num_processes": 2}, DDPStrategy, "ddp", CPUAccelerator, 0),
+        ({"accelerator": "dp", "num_processes": 2}, DDPStrategy, "ddp", CPUAccelerator, 0),
     ],
 )
-def test_trainer_config_accelerator(monkeypatch, trainer_kwargs, strategy_cls, strategy_name, _device_type, num_gpus):
+def test_trainer_config_accelerator(
+    monkeypatch, trainer_kwargs, strategy_cls, strategy_name, accelerator_cls, num_gpus
```
Comment (kaushikb11) [inline tests/trainer/test_trainer.py:1200]: Same here. There should be a follow-up PR to make the trainer config tests `gpus` agnostic.

Comment (rohitgr7) [reply tests/trainer/test_trainer.py:1200]: yeah! I think someone is making some PRs to update these.

**Review by tchaton — APPROVED** [review]
LGMT !

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (carmocca) [reply tests/trainer/test_trainer.py:1200]: Feel free to work on this follow-up!
